pyriemann/embedding.py

```diff
@@ -438,28 +446,344 @@ def locally_linear_embedding(
+        algorithm.
+
+        Parameters
+        ----------
+        Y : ndarray, shape (n_matrices, n_components, n_components)
+            Set of SPD matrices.
+
+        Returns
+        -------
+        Q : ndarray, shape (n_matrices, n_matrices)
+            The matrix of the low dimensional similarities conditional
+            probabilities of Y.
+        Dsq : ndarray, shape (n_matrices, n_matrices)
+            The array containing the squared Riemannian distances between
+            the points in X.
+        """
+        n_matrices, _, _ = Y.shape
+        Dsq = pairwise_distance(Y, squared=True)
+
+        denominator = np.sum(
+            [np.sum([np.delete(1 / (1 + Dsq[k, :]), k)])
+             for k in range(n_matrices)]
+        )
+        Q = 1 / (1 + Dsq) / denominator
+        np.fill_diagonal(Q, 0)
+        return Q, Dsq
+
+    def _cost(self, P, Q):
+        """Computed the loss of the t-SNE, that is the Kullback-Leibler
+        divergence between P and Q.
+
+        Parameters
+        ----------
+        P : ndarray, shape (n_matrices, n_matrices)
+            The matrix of the symmetrized conditional probabilities of X.
+        Q : ndarray, shape (n_matrices, n_matrices)
+            The matrix of the low dimensional similarities conditional
+            probabilities of Y.
+
+        Returns
+        -------
+        _ : float
+            The cost of the t-SNE.
+        """
+        return np.sum(P * np.log((P + np.eye(P.shape[0])) /
+                                 (Q + np.eye(P.shape[0]))))
+
+    def _riemannian_gradient(self, Y, P, Q, Dsq):
+        """Computed the Riemannian gradient of the loss of the t-SNE.
+
+        Parameters
+        ----------
+        Y : ndarray, shape (n_matrices, n_components, n_components)
+            Set of SPD matrices.
+        P : ndarray, shape (n_matrices, n_matrices)
+            The matrix of the symmetrized conditional probabilities of X.
+        Q : ndarray, shape (n_matrices, n_matrices)
+            The matrix of the low dimensional similarities conditional
+            probabilities of Y.
+        Dsq : ndarray, shape (n_matrices, n_matrices)
+            The Riemannian distance matrix of Y.
+
+        Returns
+        -------
+        grad : ndarray, shape (n_matrices, n_components, n_components)
+            The Riemannian gradient of the cost of the t-SNE.
+        """
+        n_matrices, _ = P.shape
+        grad = np.zeros((n_matrices, self.n_components, self.n_components))
+        Y_i_invsqrt = invsqrtm(Y)
+        Y_i_sqrt = sqrtm(Y)
+        for i in range(n_matrices):
+            log_riemann = (
+                Y_i_sqrt[i] @ logm(Y_i_invsqrt[i] @ Y @ Y_i_invsqrt[i])
+                @ Y_i_sqrt[i]
+            )
+            grad[i] = -4 * np.sum(
+                ((P[i] - Q[i]) / (1 + Dsq[i]))[:, np.newaxis, np.newaxis] *
+                log_riemann,
+                axis=0,
+            )
+        return grad
+
+    def _run_minimization(self, P):
```

Comment:
If you look at `TLRotate.fit()`, you can see that most of the optimization is done on the `grassmann` module.
https://github.com/pyRiemann/pyRiemann/blob/master/pyriemann/transfer/_estimators.py#L693

Can you do the same for `TNSE.fit()`, and put `_run_minimization` (and subfunctions) in module `positive_definite`?